pkg/controllers/cspi-controller/handler.go

```diff
@@ -241,6 +227,51 @@ func (c *CStorPoolInstanceController) updateStatus(cspi *cstor.CStorPoolInstance
 	return nil
 }
 
+// updateROMode sets/unsets the pool readonly mode property. It does the following changes
+// 1. If pool used space reached to roThresholdLimit then pool will be set to readonly mode
+// 2. If pool was in readonly mode if roThresholdLimit/pool expansion was happened then it
+//    unsets the ReadOnly Mode.
+// NOTE: This function must be invoked after having the updated cspiStatus information
+func (c *CStorPoolInstanceController) updateROMode(
+	cspiStatus *cstor.CStorPoolInstanceStatus, cspi cstor.CStorPoolInstance) {
+	roThresholdLimit := cspi.Spec.PoolConfig.ROThresholdLimit
+	totalInBytes := cspiStatus.Capacity.Total.Value()
+	usedInBytes := cspiStatus.Capacity.Used.Value()
+	pool := zpool.PoolName()
+
+	usedPercentage := (usedInBytes * 100) / totalInBytes
+	if (int(usedPercentage) >= roThresholdLimit) &&
+		(roThresholdLimit != 0 && roThresholdLimit != 100) {
+		if !cspiStatus.ReadOnly {
+			if err := zpool.SetPoolRDMode(pool, true); err != nil {
+				// Here, we are just logging in next reconciliation it will be retried
+				klog.Errorf("failed to set pool ReadOnly Mode to %t error: %s", true, err.Error())
+			} else {
+				cspiStatus.ReadOnly = true
+				c.recorder.Event(&cspi,
+					corev1.EventTypeWarning,
+					"PoolReadOnlyThreshold",
+					"Pool storage limit reached to thresholdLimit. Pool expansion is required to make it's volume replicas RW",
+				)
+			}
+		}
+	} else {
+		if cspiStatus.ReadOnly {
+			if err := zpool.SetPoolRDMode(pool, false); err != nil {
+				klog.Errorf("Failed to unset pool readOnly mode : %v", err)
+			} else {
+				cspiStatus.ReadOnly = false
+				c.recorder.Event(&cspi,
+					corev1.EventTypeNormal,
+					"PoolReadOnlyThreshold",
+					"Pool roThresholdLimit or pool got expanded due to that pool readOnly mode is unset",
+				)
+			}
+
+		}
+	}
+}
+
```

Comment:
Similar comment as above... 
Pool should go into RO if it needs to go and pool should go into RW if it needs to go irrespective of status...
Finally, the status can be updated with RO thing.
Happy to learn more and discuss.